File: FootballApp/api/FootballApp.Api/Models/PagedResult.cs
namespace FootballApp.Api.Models;

/// <summary>
/// Generic paged response wrapper. Reusable for any list endpoint, not
/// just football clubs.
///
/// totalCount is critical for server-side pagination: the client needs
/// it to compute pageCount = ceil(totalCount / pageSize), which drives
/// the page-number controls in TanStack Table.
/// </summary>
public sealed class PagedResult<T>
{
    public IReadOnlyList<T> Items { get; init; } = Array.Empty<T>();

    public int TotalCount { get; init; }

    public int PageIndex { get; init; }

    public int PageSize { get; init; }

    /// <summary>
    /// Convenience derived value. Computed once on serialization.
    /// </summary>
    public int PageCount =>
        PageSize <= 0 ? 0 : (int)Math.Ceiling((double)TotalCount / PageSize);

    public PagedResult() { }

    public PagedResult(IReadOnlyList<T> items, int totalCount, int pageIndex, int pageSize)
    {
        Items = items;
        TotalCount = totalCount;
        PageIndex = pageIndex;
        PageSize = pageSize;
    }
}

File: FootballApp/web/football-app-web/src/app/models/paged-result.ts
/**
 * Generic paged response wrapper. Mirrors C# PagedResult<T>.
 *
 * pageCount is server-computed (ceil(totalCount / pageSize)) so the client
 * never has to do that math itself. TanStack Table consumes it directly.
 */
export interface PagedResult<T> {
    readonly items: ReadonlyArray<T>;
    readonly totalCount: number;
    readonly pageIndex: number;
    readonly pageSize: number;
    readonly pageCount: number;
}